### PR TITLE
[SPARK-41260][PYTHON][SS] Cast NumPy instances to Python primitive types in GroupState update

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to cast NumPy instances in `GroupState.update`.  Previously, if we pass a NumPy instance to `GroupState.update`, then it failed with an exception as below:

```
net.razorvine.pickle.PickleException: expected zero arguments for construction of ClassDict (for numpy.dtype). This happens when an unsupported/unregistered class is being unpickled that requires construction arguments. Fix it by registering a custom IObjectConstructor for this class.
	at net.razorvine.pickle.objects.ClassDictConstructor.construct(ClassDictConstructor.java:23)
	at net.razorvine.pickle.Unpickler.load_reduce(Unpickler.java:759)
	at net.razorvine.pickle.Unpickler.dispatch(Unpickler.java:199)
	at net.razorvine.pickle.Unpickler.load(Unpickler.java:109)
	at net.razorvine.pickle.Unpickler.loads(Unpickler.java:122)
	at org.apache.spark.sql.api.python.PythonSQLUtils$.$anonfun$toJVMRow$1(PythonSQLUtils.scala:125)
```

### Why are the changes needed?

`applyInPandasWithState` uses pandas instances so it is very common to extract a NumPy instance from the pandas and set it to the group state.

### Does this PR introduce _any_ user-facing change?

No because this new API is not released yet.

### How was this patch tested?

Manually tested, and unittest was added.

### CODE REVIEW

File: python/pyspark/sql/streaming/state.py [hunk 1]
```diff
@@ -130,6 +131,10 @@ def update(self, newValue: Tuple) -> None:
         if newValue is None:
             raise ValueError("'None' is not a valid state value")
 
+        if has_numpy:
+            import numpy as np
+            # In order to convert NumPy types to Python primitive types.
+            newValue = tuple(v.tolist() for v in newValue if isinstance(v, np.generic))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming/state.py:137]: @xinrong-meng mind reviewing this please?

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR FYI

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsInPandasWithStateSuite.scala [hunk 2]
```diff
@@ -803,4 +803,49 @@ class FlatMapGroupsInPandasWithStateSuite extends StateStoreMetricsTest {
...
+      |import pandas as pd
+      |import numpy as np
+      |from pyspark.sql.types import StructType, StructField, StringType
+      |
+      |tpe = StructType([
+      |    StructField("key", StringType()),
+      |    StructField("valueAsString", StringType())])
+      |
+      |def func(key, pdf_iter, state):
+      |    np_value = np.int64(1)  # NumPy instance
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsInPandasWithStateSuite.scala:821]: Could we please imagine/check some more types which is used widely and add them to the test here as well? No need to be exhaustive but single check seems to be too specific.

**Review by HeartSaVioR — APPROVED** [review]
+1 I can see the test checks the (implicit) type conversions properly. Nice addition!

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Comment by HyukjinKwon** [comment]
Thanks!

**Pull request closed without merging**